### DISCUSSION TITLE
Sprite won't be visible until the next sprite spawns

### DISCUSSION THREAD

**Discussion by nskins** [bevyengine/bevy#8288] [Q&A] [answered]
Hi! I'm running into a strange issue where the sprite will spawn "invisibly" and then only become visible once the next sprite spawns. I have code to reproduce this and have included a video. I am very much a beginner with Bevy, so I wouldn't know if this is a bug with my code or an issue with the engine.

In the video, watch when the score goes from 160 -> 165. You'll see that a banana appears to materialize in the middle of the screen. But it had spawned at the top of the screen like the rest of the bananas.

I would very much appreciate to hear any thoughts about this issue. Please let me know if I can be of any further assistance!

- Code example: https://github.com/nskins/banana_catch (branch: `main`)
- Relevant issue: https://github.com/nskins/banana_catch/issues/6
- Additional discussion: https://github.com/nskins/banana_catch/pull/5#pullrequestreview-1362103442

[Screencast from 03-28-2023 08:53:47 PM.webm](https://user-images.githubusercontent.com/7865228/228398925-90804706-0b79-42e6-a630-00147bdb01d6.webm)

**Comment by guest-twenty-one** [comment, marked as answer]
This looks like an asset unloading bug:

1. at `0:14` the last single banana is caught and its entity is despawned
2. the asset server marks the `banana.png` image for removal as there's no sprite using it
3. **ONE** or **TWO** frames later, another single banana sprite is spawned using the same `banana.png` image
4. the asset server unloads the `banana.png` image (even though it shouldn't), and your banana sprite disappears from the screen (the entity is still there and moving down - you just don't see it)
5. it appears after another newly spawned sprite loads the `banana.png` image again

What's interesting is that, if the new sprite is spawned in the same frame or 3 or more frames after the last banana sprite is despawned (step 1), then everything works fine.

Anyway, you should cache your images/assets to stop them from unloading:

```rust
fn main() {
    App::new()
        // ...
        .add_startup_system(load_and_cache_images)
        // ...
}

#[derive(Resource)]
struct ImageCache {
	banana: Handle<Image>,
	bunch_of_bananas: Handle<Image>
}

fn load_and_cache_images(
    mut commands: Commands,
    asset_server: Res<AssetServer>,
) {
    commands.insert_resource(ImageCache {
        banana: asset_server.load("sprites/banana.png"),
        bunch_of_bananas: asset_server.load("sprites/bananabunch.png")
    });
}
```

then use the cached images when spawning sprites:

```rust
fn spawn_falling_objects_over_time(
    mut commands: Commands,
    window_query: Query<&Window, With<PrimaryWindow>>,
    falling_object_spawn_timer: Res<FallingObjectSpawnTimer>,
    image_cache: Res<ImageCache>,
) {
    if falling_object_spawn_timer.timer.finished() {
        let window = window_query.single();

        let bounds_width = window.width() - (2.0 * BOUND_SIZE);

        let random_x = (random::<f32>() * bounds_width) + BOUND_SIZE;

        let (kind, points, texture) = if random::<f32>() < 0.1 {
            (FallingObjectKind::BananaBunch, 5, image_cache.bunch_of_bananas.clone_weak())
        } else {
            (FallingObjectKind::Banana, 1, image_cache.banana.clone_weak())
        };	
		
        commands.spawn((
            SpriteBundle {
                transform: Transform::from_xyz(random_x, window.height(), 0.0),
                texture,
                ..default()
            },
            FallingObject {
                kind,
                points,
            },
        ));
    }
}
```
edit: This seems to be the same bug as: #6626

**Reply by nskins** [reply]
Hi @guest-twenty-one! Thank you so much for taking a look at my issue and providing such a detailed response. I appreciate the insight you've given me into how Bevy handles assets. 

To add on to this, reading the [Handle](https://docs.rs/bevy/latest/bevy/asset/struct.Handle.html) documentation was very informative. The fix above will keep a strong handle available and thus will not allow the assets to unload. Just want to throw that out there in case anyone else runs into this problem. 🙂
